app/controllers/submissions_controller.rb

```diff
@@ -202,6 +240,111 @@ def downloadAll
               filename: "#{@course.name}_#{@course.semester}_#{@assessment.name}_submissions.zip")
   end
 
+  action_auth_level :downloadBatch, :course_assistant
+  def downloadBatch
+    submission_ids = params[:submission_ids]
+    flash[:error] = "Cannot index submissions for nil assessment" if @assessment.nil?
+
+    unless @assessment.valid?
+      @assessment.errors.full_messages.each do |msg|
+        flash[:error] += "<br>#{msg}"
+      end
+      flash[:html_safe] = true
+    end
+
+    if @assessment.disable_handins
+      flash[:error] = "There are no submissions to download."
+      if @cud.course_assistant
+        redirect_to([@course, @assessment])
+      else
+        redirect_to([@course, @assessment, :submissions])
+      end
+      return
+    end
+
+    submissions = submission_ids.map { |sid| @assessment.submissions.find_by(id: sid) }
+    submissions = submissions.select { |s| s && @cud.can_administer?(s.course_user_datum) }
+
+    filedata = submissions.collect do |s|
+      p = s.handin_file_path
+      email = s.course_user_datum.user.email
+      [p, download_filename(p, email)] if !p.nil? && File.exist?(p) && File.readable?(p)
+    end.compact
+
+    result = Archive.create_zip filedata # result is stringIO to be sent
+    if result.nil?
+      flash[:error] = "There are no submissions to download."
+      if @cud.course_assistant
+        redirect_to([@course, @assessment])
+      else
+        redirect_to([@course, @assessment, :submissions])
+      end
+      return
+    end
+
+    send_data(result.read, # to read from stringIO object returned by create_zip
+              type: "application/zip",
+              disposition: "attachment", # tell browser to download
+              filename: "#{@course.name}_#{@course.semester}_#{@assessment.name}_submissions.zip")
+  end
+
+  action_auth_level :excuseBatch, :course_assistant
+  def excuseBatch
+    submission_ids = params[:submission_ids]
+    flash[:error] = "Cannot index submissions for nil assessment" if @assessment.nil?
+
+    unless @assessment.valid?
+      @assessment.errors.full_messages.each do |msg|
+        flash[:error] += "<br>#{msg}"
+      end
+      flash[:html_safe] = true
+    end
+
+    submissions = submission_ids.map { |sid| @assessment.submissions.find_by(id: sid) }
+
+    submissions.each do |submission|
+      next if submission.nil?
```

Comment:
Theoretically there would not be case it is nil, but an instructor or admin could also manually add nil entries (not sure why anyone would do this but still possible nonetheless) to the database and it would result in nil entries for submissions. I see there are nil checks throughout the file so I think this is fine for sanity sake.